Packet++/src/Asn1Codec.cpp

```diff
@@ -0,0 +1,569 @@
+	{
+		switch (m_ValueLength)
+		{
+			case 1:
+			{
+				m_Value = *(uint8_t*)data;
+				break;
+			}
+			case 2:
+			{
+				m_Value = be16toh(*(uint16_t*)data);
+				break;
+			}
+			case 3:
+			{
+				uint8_t tempData[4] = {0};
+				memcpy(tempData + 1, data, 3);
+				m_Value = be32toh(*(uint32_t*)tempData);
+				break;
+			}
+			case 4:
+			{
+				m_Value = be32toh(*(uint32_t*)data);
+				break;
+			}
+			default:
+			{
+				throw std::runtime_error("An integer ASN.1 record of more than 4 bytes is not supported");
+			}
+		}
+	}
+
+	std::vector<uint8_t> Asn1IntegerRecord::encodeValue() const
+	{
+		std::vector<uint8_t> result;
+		result.reserve(m_ValueLength);
+
+		switch (m_ValueLength)
+		{
+			case 1:
+			{
+				result.push_back(static_cast<uint8_t>(m_Value));
+				break;
+			}
+			case 2:
+			{
+				uint8_t tempArr[sizeof(uint16_t)];
+				auto hostValue = htobe16(static_cast<uint16_t>(m_Value));
+				memcpy(tempArr, &hostValue, m_ValueLength);
+				std::copy(tempArr, tempArr + m_ValueLength, std::back_inserter(result));
+				break;
+			}
+			case 3:
+			{
+				uint8_t tempArr[sizeof(uint32_t)];
+				auto hostValue = htobe32(static_cast<uint32_t>(m_Value));
+				memcpy(tempArr, &hostValue, sizeof(uint32_t));
+				std::copy(tempArr + 1, tempArr + m_ValueLength + 1, std::back_inserter(result));
+				break;
+			}
+			case 4:
+			{
+				uint8_t tempArr[sizeof(uint32_t)];
+				auto hostValue = htobe32(static_cast<uint32_t>(m_Value));
+				memcpy(tempArr, &hostValue, m_ValueLength);
+				std::copy(tempArr, tempArr + m_ValueLength, std::back_inserter(result));
+				break;
+			}
+			default:
+			{
+				throw std::runtime_error("Integer value of more than 4 bytes is not supported");
+			}
+		}
+
+		return result;
+	}
+
+	Asn1EnumeratedRecord::Asn1EnumeratedRecord(uint32_t value) : Asn1IntegerRecord(value)
+	{
+		m_TagType = static_cast<uint8_t>(Asn1UniversalTagType::Enumerated);
+	}
+
+	Asn1OctetStringRecord::Asn1OctetStringRecord(const std::string& value) : Asn1PrimitiveRecord(static_cast<uint8_t>(Asn1UniversalTagType::OctetString))
+	{
+		m_Value = value;
+		m_ValueLength = value.size();
+		m_TotalLength = m_ValueLength + 2;
+	}
+
+	void Asn1OctetStringRecord::decodeValue(uint8_t* data, bool lazy)
+	{
+		m_Value = std::string(reinterpret_cast<char*>(data), m_ValueLength);
+	}
+
+	std::vector<uint8_t> Asn1OctetStringRecord::encodeValue() const
+	{
+		return {m_Value.begin(), m_Value.end()};
+	}
+
+	Asn1BooleanRecord::Asn1BooleanRecord(bool value) : Asn1PrimitiveRecord(static_cast<uint8_t>(Asn1UniversalTagType::Boolean))
+	{
+		m_Value = value;
+		m_ValueLength = 1;
+		m_TotalLength = 3;
+	}
+
+	void Asn1BooleanRecord::decodeValue(uint8_t* data, bool lazy)
+	{
+		m_Value = data[0] != 0;
+	}
+
+	std::vector<uint8_t> Asn1BooleanRecord::encodeValue() const
+	{
+		uint8_t byte = (m_Value ? 0xff : 0x00);
+		return { byte };
+	}
+
+	Asn1NullRecord::Asn1NullRecord() : Asn1PrimitiveRecord(static_cast<uint8_t>(Asn1UniversalTagType::Null))
```

Comment:
why not have a overloading for `Asn1PrimitiveRecord` to use `Asn1UniversalTagType` as the argument?

Reply:
Do you mean create another c'tor for `Asn1PrimitiveRecord` that takes an argument of type `Asn1UniversalTagType`?

Reply:
yes `Asn1PrimitiveRecord(Asn1UniversalTagType type)`, I think it is clearer.

Reply:
Fixed in this commit: https://github.com/seladb/PcapPlusPlus/pull/1353/commits/4b4865f2aa01e3c5ac619f91d7a6d0ff0227934c